backend/src/models/ScenarioGraph.js

```diff
@@ -0,0 +1,123 @@
+/**
+ * Creates a graph representation of a list of scenes.
+ * @param scene - An array of scene objects
+ */
+export default class ScenarioGraph {
+
+    // Store graph as a modified adjacency list for easy indexing
+    graph = {};
+
+    constructor(scenes) {
+
+        // Iterate over each scene
+        scenes.forEach((scene) => {
+
+            let children = [];
+            let leaf = true;
+
+            // Iterate over each component of a scene
+            scene.components.forEach((component) => {
+                
+                // Check if the scene has any links to other scenes
+                if (component.type === 'BUTTON') {
+                    children.push(component.nextScene);
+                    leaf = false;    // Scene is not a leaf as it has a child.
+                }
+            })

```

Comment:
Some JS syntactical sugar with functional programming if you want to use it to simplify your code :P 

```
                  
const children = scene.components
                  .filter((it) => it.type === 'BUTTON'))
                  .map((it) => it.nextScene));
// If the current scene has no children, then by definition it is a leaf node
const leaf = children.length === 0;